bubuku/features/restart_if_dead.py

```diff
@@ -14,22 +15,38 @@ def __init__(self, broker: BrokerManager, zk: BukuExhibitor):
         self.broker = broker
         self.zk = zk
         self.need_check = True
+        self.last_zk_session_check = datetime.now()
 
     def check(self) -> Change:
         if not self.need_check:
             return None
-        if self.broker.is_running_and_registered():
+        if self.is_running_and_registered(should_retry=True):
             return None
 
-        _LOG.info('Oops! Broker is dead, triggering restart')
+        _LOG.warning('Oops! Broker is dead, triggering restart')
         self.need_check = False
 
         # Do not start if broker is running and registered
         def _cancel_if():
-            return self.broker.is_running_and_registered()
+            return self.is_running_and_registered()
 
         return RestartBrokerChange(self.zk, self.broker, _cancel_if, self.on_check_removed)
 
+    # Attempt to verify that broker is not registered in zookeeper for twice as long as the zookeeper session timeout.
+    # Allow zookeeper client to try to restore the session before killing tha kafka process as soon as zookeeper session is dead.
+    def is_running_and_registered(self, should_retry=False):
+        if not self.broker.is_running():
+            return False
+        if not self.broker.is_registered_in_zookeeper():
+            _LOG.warning('Broker is to not be regiestered in Zookeeper')
+            time_to_return = self.last_zk_session_check + timedelta(milliseconds=self.broker.get_zookeeper_session_timeout() * 2)
+            if datetime.now() > time_to_return or not should_retry:
```

Comment:
2 things:
- I would check for `should_retry` before doing any calculations
- I would extract datetime.now(). into variable, like current_time = datetime.now(), since it is used 3 times